### PR TITLE
Features API

### PR DESCRIPTION
**Pull request by thecatcore**
A lot of things to do.

### CODE REVIEW

**Comment by thecatcore** [comment]
Nice feature mojang : [MC-194350](https://bugs.mojang.com/browse/MC-194350)

**Comment by thecatcore** [comment]
It's ready for review but I prefer to wait until tomorrow snapshot as it will eventually break it.

File: src/main/java/com/swordglowsblue/artifice/api/builder/data/StateDataBuilder.java [hunk 1]
```diff
@@ -3,11 +3,11 @@
 import com.google.gson.JsonObject;
 import com.swordglowsblue.artifice.api.builder.TypedJsonBuilder;
 
-public class BlockStateDataBuilder extends TypedJsonBuilder<JsonObject> {
```
Comment (natanfudge) [inline src/main/java/com/swordglowsblue/artifice/api/builder/data/StateDataBuilder.java:6]: is this class something you added recently?

**Review by natanfudge — COMMENTED** [review]
wow that's a lot of stuff

File: src/main/java/com/swordglowsblue/artifice/api/builder/data/StateDataBuilder.java (hunk 1 above)
Comment (thecatcore) [reply src/main/java/com/swordglowsblue/artifice/api/builder/data/StateDataBuilder.java:6]: yeah I think :/ .
It's also used by FluidState so I renamed it StateDataBuilder.

**Pull request merged**